README.md
Welcome into the Jarr Project


This is our school robotic project.
The aim is to controle the Ar Drone with moving our body.


Co workers : 
Amalric de Buffières (Therailwayking)
Julien Perret 
Raphaël Haennig 
Raphaël Prevost (MaitreCourgette)






For this project, we use the python-ardrone :
https://github.com/venthur/python-ardrone



Changelog.txt
27.05.2016 :
We created a new Github repository (The old one was so messy)


main.py
#Python script

import traitement_image as Ti
import drone_relation as DroneSimulated
import libardrone
import mouvement as Move

from PIL import Image as Image



def main():
    # drone = Drone.Drone()
    # drone = libardrone.ARDrone()



    """
    move = Move.Mouvement()
    move.initDrone()



    counter = 1

    while counter < 3 :
        print ""
        print ""
        print ""
        print "Starting new thing"
        image = DroneSimulated.getImage(counter)
        zones = Ti.manageImage(image)
        move.getMove(zones)

        #if move == -1:
         #   break

        counter += 1
        print "Counter = " + str(counter)


    move.makeMove()

    while (raw_input("Type anything")):
        print "test"




    Ti.stopDrone()

    """



    zones = Ti.manageImage(Image.open("Images/ImageBigTest2.png"))

    for zone in zones:
        outputString = ""

        outputString += str(zone.x) + " "
        outputString  += str(zone.y) + " "
        outputString += str(zone.side)

        print outputString

    drone = Drone.Drone()
    drone.speed = 0.1
    while True : 
        image = Drone.getImage()
        zones = Ti.manageImage(image)
        move = Move.getMove(zones)


main()









controle.py
# -*- coding: utf-8 -*-
from numpy import tan
from numpy import pi

class Drone :
    
    def __init__(self):
        self.launch()

    def launch(self):
        self.setHeight(1.5)
    
    def setHeight(self, height):
        self.height=height
    
    def rotateSelf(self):
        return True
        
    def rotateLeft(self, distance):
        return True

    def rotateRight(self, distance):
        return True

    def leave(self):
        return True

    def comeBack(self):
        return True

    def left(self):
        return True

    def right(self):
        return True
    
    def land(self):
        print "J'atteris"

class Barre :
    
    def __init__(self, x, y, side):
        self.x=x
        self.y=y
        self.side=side
        print("Creation d'une barre !")
    
    def decrisToi(self, string):
        print("La barre {} a pour coordonnees ({},{}).\n".format(string, self.x, self.y))
        


def main():

        drone = Drone()
        b1 = Barre(100,250,True)
        b0 = Barre(360,250,False)
        b2 = Barre(500,250,True)
        
        tab = [b1,b0,b2]
        
        for i in range(len(tab)):
            
            if tab[i].side == False or len(tab) == 1 :
                barreMilieu = tab[i]
                barreMilieu.decrisToi("du milieu")
                
            else :
                if i > 0 :

                    if tab[i].x > tab[i-1].x :
                        barreDroite = tab[i]
                        barreDroite.decrisToi("de droite")
                    else :
                        barreGauche = tab[i]
                        barreGauche.decrisToi("de gauche")

                else :
                    if tab[i].x > tab[i+1].x :
                        barreDroite = tab[i]
                        barreDroite.decrisToi("de droite")
                    else :
                        barreGauche = tab[i]
                        barreGauche.decrisToi("de gauche")
        
        distanceDroneHomme = 5
        
        pixelHoriBGBD = abs(barreDroite.x-barreGauche.x)
        distanceHoriBGBD = distanceDroneHomme*tan(93*(pi/180)*pixelHoriBGBD/640)
        
        pixelHoriBGBM = abs(barreGauche.x-barreMilieu.x)
        distanceHoriBGBM = distanceDroneHomme*tan(93*(pi/180)*pixelHoriBGBM/640)
        
        pixelHoriBMBD = abs(barreMilieu.x-barreDroite.x)
        distanceHoriBMBD = distanceDroneHomme*tan(93*(pi/180)*pixelHoriBMBD/640)         
           
        pixelVertiBGBD = abs(barreDroite.y-barreGauche.y)
        distanceVertiBGBD = distanceDroneHomme*tan(93*(pi/180)*pixelVertiBGBD/640)
        
        pixelVertiBGBM = abs(barreGauche.y-barreMilieu.y)
        distanceVertiBGBM = distanceDroneHomme*tan(93*(pi/180)*pixelVertiBGBM/640)
        
        pixelVertiBMBD = abs(barreMilieu.y-barreDroite.y)
        distanceVertiBMBD = distanceDroneHomme*tan(93*(pi/180)*pixelVertiBMBD/640)
        
        pixelHoriBMOr = barreMilieu.x-320
        
        if pixelHoriBMOr < 0 :
            distanceHoriBMOr = distanceDroneHomme*tan(93*(pi/180)*pixelHoriBMOr/640)
        elif pixelHoriBMOr > 0 :
            distanceHoriBMOr = -distanceDroneHomme*tan(93*(pi/180)*pixelHoriBMOr/640)

        if len(tab) <= 1 :
            drone.rotateSelf()

        elif len(tab) == 2 :
            if barreMilieu == None :
                return True
            elif barreGauche == None :
                drone.rotateRight()
            else : 
                drone.rotateGauche()
        
        else :
            
            if distanceHoriBGBM < 0.05 and distanceHoriBMBD < 0.05 :
                drone.land()
                
            elif 0.2 < distanceVertiBGBM < 0.9 or 0.2 < distanceVertiBMBD < 0.9 :
                drone.setAltitude(19.2857*distanceVertiBGBM - 2.3571)
                
            elif pixelHoriBGBD < 40 :
                while pixelHoriBGBD < 40 :
                    drone.comeBack()
                    
            elif pixelHoriBGBD > 80 :
                while pixelHoriBGBD > 80 :
                    drone.leave()
            
            elif distanceHoriBMOr < 0.3 :
                drone.left()
            
            elif barreMilieu.x > 6 :
                drone.right()
            
            
            
            
if __name__ == '__main__' :
    main()

drone_relation.py
# -*- coding: utf-8 -*-
"""
Created on Fri Apr 15 10:38:16 2016

@author: julienperret
"""

from PIL import Image as Image



def getImage(testProcedureImageValue):
    print "Getting the image"
    # drone = Drone()

    fileName = "Images/ImageBigTest" + str(testProcedureImageValue) + ".png"

    # fileName = "Images/ImageBigTest5.png"
    img = Image.open(fileName)


    img = Image.open("ImageBigTest.png")
    return img

class Drone :


    def __init__(self):
        self.takeoff()

    def takeoff(self):
        print "I'm taking off"
        self.setHeight(1.5)

    def setHeight(self, height):
        self.height=height
        print "I'm going to ",height,"m"

    def rotateSelf(self):
        print "I'm rotating on myself"
        return True

    def rotateLeft(self):
        print "I'm rotating left"
        self.turn_left()
        return True

    def rotateRight(self):
        print "I'm rotating right"
        self.turn_right()
        return True

    def move_backward(self):
        print "I'm moving backward"
        return True

    def move_forward(self):
        print "I'm moving forward"
        print

    def left(self):
        print "I'm going left"
        return True

    def right(self):
        print "I'm going right"
        return True

    def land(self):
        print "I'm landing"
        self.setHeight(0)

    def hover(self):
        print "I'm hovering"
        return True

    def halt(self):
        print "I'm halting"
        return True

    def turn_left(self):
        print "I'm turning left"
        return True

    def turn_right(self):
        print "I'm turning right"
        return True


controle_clavier.py
# -*- coding: utf-8 -*-
"""
Created on Mon Apr 11 09:49:54 2016

@author: julienperret
"""


class Drone :
    
    def __init__(self):
        self.takeoff()

    def takeoff(self):
        self.setHeight(1.5)
    
    def setHeight(self, height):
        self.height=height
    
    def rotateSelf(self):
        return True
        
    def rotateLeft(self, distance):
        return True

    def rotateRight(self, distance):
        return True

    def move_backward(self):
        return True

    def move_forward(self):
        print 

    def left(self):
        return True

    def right(self):
        return True
    
    def land(self):
        self.setHeight(0)
    
    def hover(self):
        return True
    
    def halt(self):
        return True
    
    def turn_left(self):
        return True
    
    def turn_right(self):
        return True
        
def main():
    
    drone = Drone()
    
    v = input("Choisissez moi une vitesse de deplacement : ")

    while 1==1 :
        
        saisie = raw_input("Saisissez une commande : ")
        
        if saisie == "up" :
            drone.takeoff()
            print "Je décolle\n"
        
        elif saisie == "down" :
            drone.land()
            print "J'atteris\n"
            
        elif saisie == "moveforward":
            drone.speed=v
            drone.move_forward()
            print "J'avance\n"
        
        elif saisie == "movebackward":
            drone.speed=v
            drone.move_backward()
            print "Je recule\n"
            
        elif saisie == "moveright" :
            drone.move_right()
            print "Je me deplace a droite\n"
        
        elif saisie == "moveleft" :
            drone.move_left()
            print "Je me deplace a gauche\n"
        
        elif saisie == "turnleft" :
            drone.turn_left()
            print "Je tourne sur ma gauche"

        elif saisie == "turnright" :
            drone.turn_right()
            print "Je tourne sur ma droite"
            
        elif saisie == "hover" :
            drone.hover()
            print "Je stagne\n"
            
        else :
            break

    drone.land()
    print "J'atteris"
    drone.halt()

if __name__ == '__main__' :
    main()

traitement_image.py
# -*- coding: utf-8 -*-
"""
Created on Fri Apr 15 10:39:25 2016

@author: julienperret
"""

from scipy import *
from PIL import Image as Image


## a pixel is a image basic element. Can determinate if he is red and
## know its PixelGroup
class Pixel:
    # We are searching for red
    minRedValue = 190
    maxOtherValues = 50

    test = 0;



    def __init__(self, x, y, r, g, b):
        self.x = x
        self.y = y
        self.r = r
        self.g = g
        self.b = b
        self.pixelGroup = False

    def displayThisPixel(self):  # I hate it, but it's useful sometimes
        print "This pixel position is x = " + str(self.x) + " y = " + str(self.y) + " and its color is " + str(
            self.r) + "." + str(self.g) + "." + str(self.b)

    def quickDisplay(self):

        outputString = "(" + str(self.x) + "," + str(self.y) + " [" + str(self.r) + "," + str(self.g) + "," + str(
            self.b) + "]"
        if self.pixelGroup != False:
            outputString += str(self.pixelGroup.id) + ","
        else:
            outputString += "N,"

        if self.isRed():
            outputString += "R)"
        else:
            outputString += "O)"

        for it in range(len(outputString), 23):
            outputString += " "
        return outputString

    def isRed(self):
        if self.r > Pixel.minRedValue and self.g < Pixel.maxOtherValues and self.b < Pixel.maxOtherValues:
            return True
            # print "I'm Red"
        else:
            return False
            # print "I'm not red"

    def setPixelGroup(self, pixelGroup):
        self.pixelGroup = pixelGroup

    def belongToPixelGroup(self):  # if the pixel belong to a pixel group
        if self.pixelGroup != False:
            return True
        else:
            return False




## A PixelGroup is a neighborhood of pixel on the image
## Has average X and Y coordinates, the amount of pixel he owns and his ID number
class PixelGroup:
    actualAmount = 0

    def __init__(self):
        self.id = PixelGroup.actualAmount + 1
        PixelGroup.actualAmount += 1
        self.pixelsList = []
        self.side = False
        self.x = 0
        self.y = 0

    def addPixel(self, pixel):
        self.pixelsList.append(pixel)

    def calculateAverageCoordinates(self):
        # print "id = " + str(self.id) + " pixelsList size = " + str(len(self.pixelsList))

        self.x = 0
        self.y = 0
        for pixel in self.pixelsList:
            self.x = self.x + pixel.x
            self.y = self.y + pixel.y
        if len(self.pixelsList) > 0:
            self.x /= len(self.pixelsList)
            self.y /= len(self.pixelsList)

        return self.x, self.y

    def displayAllPixels(self):
        for currentPixel in self.pixelsList:
            currentPixel.displayThisPixel()

    def getAllPixels(self):
        return self.pixelsList

    def decrisToi(self, string):
        print("La barre {} a pour coordonnees ({},{}).\n".format(string, self.x, self.y))


## Creating the pixel matrix
def geratePixelMatrix(height):
    generatedPixelMatrix = []
    for it in range(height):
        generatedPixelMatrix.append([])

    return generatedPixelMatrix


## Filling matrix width pixels
def fillPixelMatrix(emptyMatrix, pointMatrix, width, height):
    for jt in range(height):
        for it in range(width):
            # print "Coordinates in the map : it : " + str(it) + " , jt : " + str(jt)

            # creating new pixel with all datas (x, y, (r, g, b))
            tempPixel = Pixel(it, jt, pointMatrix[it, jt][0], pointMatrix[it, jt][1], pointMatrix[it, jt][2])
            emptyMatrix[jt].append(tempPixel)  # adding pixel to pixel matrix
    return emptyMatrix


def displayPixelMatrix(pixelMatrix, width, height):
    for bt in range(height):  # displaying the matrix with grave
        lineString = ""
        for at in range(width):
            lineString += str(pixelMatrix[bt][at].quickDisplay()) + " "
        print lineString


def createPixelsZone(pixelMatrix, width, height):
    # all the pixels zones
    pixelsGroupsList = []

    for jt in range(height):
        for it in range(width):
            # print "Working on  it = " + str(it) + " jt = " + str(jt) + " pixel = " + str(pixelMatrix[jt][it].quickDisplay())
            # pixelMatrix[jt][it].displayThisPixel()

            if pixelMatrix[jt][it].isRed() and not pixelMatrix[jt][it].belongToPixelGroup():
                # print "This pixel is red !"

                pixelMatrix, pixelsGroupsList = infectRedZone(pixelMatrix, width, height, it, jt, pixelsGroupsList)

    return pixelMatrix, pixelsGroupsList


def infectRedZone(pixelMatrix, width, height, initialPixelX, initialPixelY, pixelsGroupsList):
    # print "Launching a red Pixel Infection from " + pixelMatrix[initialPixelY][initialPixelX].quickDisplay()

    # List with all already infected pixel
    infectedPixels = []

    pixelGroup = PixelGroup()  # Creating the pixelGroup
    pixelsGroupsList.append(pixelGroup)
    pixelMatrix[initialPixelY][initialPixelX].pixelGroup = pixelGroup

    infectedPixels.append(pixelMatrix[initialPixelY][initialPixelX])

    # print "Group " + str(pixelGroup.id) + " is created"

    # Running to every pixels to prepare infection
    for currentPixel in infectedPixels:
        # The max function avoid IndexOutOfBoundException
        for jt in range(max([currentPixel.y - 1, 0]), min([currentPixel.y + 2, height])):
            for it in range(max([currentPixel.x - 1, 0]), min([currentPixel.x + 2, width])):
                # print "Running throw pixels : " + pixelMatrix[jt][it].quickDisplay()
                if pixelMatrix[jt][it].isRed() and not pixelMatrix[jt][it].belongToPixelGroup():
                    pixelMatrix[jt][it].pixelGroup = pixelGroup
                    pixelGroup.addPixel(pixelMatrix[jt][it])

                    infectedPixels.append(pixelMatrix[jt][it])

    return pixelMatrix, pixelsGroupsList


def displayPixelsGroups(pixelsGroupsList):
    for pixelGroup in pixelsGroupsList:
        x, y = pixelGroup.calculateAverageCoordinates()
        print "x = " + str(x) + " y = " + str(y)


def getPixelsGroupsSideValue(pixelsGroups, width):

    for it in range(len(pixelsGroups)):
        pixelsGroups[it].calculateAverageCoordinates()


    if len(pixelsGroups) == 3:
        for it in range(len(pixelsGroups)):
            if pixelsGroups[it].x == max(pixelsGroups[0].x, pixelsGroups[1].x, pixelsGroups[2].x) or pixelsGroups[it].x == min(pixelsGroups[0].x, pixelsGroups[1].x, pixelsGroups[2].x):
                pixelsGroups[it].side = True
            else:
                pixelsGroups[it].side = False

    # for this search we determine the absolute value between the group and the center of the image
    # The middle group is the one that as the smallest absolute value
    if len(pixelsGroups) == 2:
        print "There is two pixels groups"
        absoluteDistance0 = getAbsoluteDistance(pixelsGroups[0].x, width);
        absoluteDistance1 = getAbsoluteDistance(pixelsGroups[1].x, width);

        pixelsGroups[0].side = True if absoluteDistance0 < absoluteDistance1 else False
        pixelsGroups[1].side = not pixelsGroups[0].side





def getAbsoluteDistance(x, imageWidth):
    return abs( x - imageWidth / 2 )











    return pixelsGroups


def manageImage(image):
    print "Managing image"

    basePixelMatrix = image.load()  # getting image as an array
    width, height = image.size  # Getting the future matrix size

    # print "This image's size is " + str(width) + " x " + str(height)

    emptyPixelMatrix = geratePixelMatrix(height)

    pixelMatrix = fillPixelMatrix(emptyPixelMatrix, basePixelMatrix, width, height)

    # displayPixelMatrix(pixelMatrix, width, height)

    pixelMatrix, pixelsGroups = createPixelsZone(pixelMatrix, width, height)



    pixelsGroups = getPixelsGroupsSideValue(pixelsGroups, width)

    return pixelsGroups

    # displayPixelMatrix(pixelMatrix, width, height)
    # displayPixelsGroups(pixelsGroups)


FirstTryWithDrone.py
#!/usr/bin/env python
import time, libardrone

drone = libardrone.ARDrone()

def run():
    v = input("Choisissez moi une vitesse de deplacement : ")

    while 1==1 :
        
        saisie = raw_input("Saisissez une commande : ")
        
        if saisie == "up" :
            drone.takeoff()
            print "Je decolle\n"
        
        elif saisie == "down" :
            drone.land()
            print "J'atteris\n"
            
        elif saisie == "moveforward":
            drone.speed=v
            drone.move_forward()
            print "J avance\n"
        
        elif saisie == "movebackward":
            drone.speed=v
            drone.move_backward()
            print "Je recule\n"
            
        elif saisie == "moveright" :
            drone.move_right()
            print "Je me deplace a droite\n"
        
        elif saisie == "moveleft" :
            drone.move_left()
            print "Je me deplace a gauche\n"
        
        elif saisie == "turnleft" :
            drone.turn_left()
            print "Je tourne sur ma gauche"

        elif saisie == "turnright" :
            drone.turn_right()
            print "Je tourne sur ma droite"
            
        elif saisie == "hover" :
            drone.hover()
            print "Je stagne\n"
            
        else :
            break

    drone.land()
    print "J'atteris"
    drone.halt()

run()

drone.land()

mouvement.py
# -*- coding: utf-8 -*-
"""
Created on Fri Apr 15 10:39:42 2016

@author: julienperret
"""
import drone_relation as DroneManagment
import traitement_image as Ti
import libardrone

from numpy import tan
from numpy import pi



#drone = libardrone.ARDrone()

class Mouvement :
    # drone = libardrone.ARDrone()

    def __init__(self):
        self.drone = libardrone.ARDrone()

        bat = self.drone.navdata.get(0, dict()).get('battery', 0)
        print "Current Battery Level : %i / 100" % bat

    def takeoff(self):
        self.drone.takeoff()
        print "I'm taking off"

    def setHeight(self, height):
        self.height = height
        print "I'm going to ", height, "m"

    def rotateSelf(self):
        self.drone.turn_left()
        print "I'm rotating on myself"

    def rotateLeft(self):
        self.drone.turn_left()
        print "I'm rotating left"

    def rotateRight(self):
        print "I'm rotating right"
        self.drone.turn_right()

    def move_backward(self):
        print "I'm moving backward"

    def move_forward(self):
        self.drone.move_forward()
        print "I'm moving forward"

    def left(self):
        self.drone.move_left()
        print "I'm going left"

    def right(self):
        self.drone.move_right()
        print "I'm going right"

    def land(self):
        self.drone.land()
        print "I'm landing"
        self.setHeight(0)

    def hover(self):
        self.drone.hover()
        print "I'm hovering"

    def halt(self):
        self.drone.halt()
        print "I'm halting"

    def turn_left(self):
        self.drone.turn_left()
        print "I'm turning left"

    def turn_right(self):
        self.drone.turn_right()
        print "I'm turning right"

    def initDrone(self):
        self.drone.speed = 0.2
        self.drone.takeoff()

    def stopDrone(self):
        self.drone.land()
        self.drone.halt()

    def makeMove(self):
        self.drone.move_right()



    def getMove(self, zones):
        print "Finding the right move !"

        distanceDroneHomme = 1

        # drone = DroneManagment.Drone()
        # drone = libardrone.ARDrone()

        # if drone < 0:
        #   print "Drone Error"
        #  return -1


        (barreGauche, barreMilieu, barreDroite) = getBarsName(zones)

        (pixelHoriBGBD,
         distanceHoriBGBD,
         pixelHoriBGBM,
         distanceHoriBGBM,
         pixelHoriBMBD,
         distanceHoriBMBD,
         pixelVertiBGBD,
         distanceVertiBGBD,
         pixelVertiBGBM,
         distanceVertiBGBM,
         pixelVertiBMBD,
         distanceVertiBMBD,
         pixelHoriBMOr,
         distanceHoriBMOr) = getMeasures(barreGauche, barreMilieu, barreDroite)

        if len(zones) <= 1:
            print "turn_left"
            self.turn_left()
        elif len(zones) == 2:
            if barreMilieu == None:
                return True
            elif barreGauche == None:
                print "Turn_right"
                self.turn_right()
            else:
                print "TurnLeft"
                self.turn_left()

        else:

            if distanceHoriBGBM < 0.05 and distanceHoriBMBD < 0.05:
                print "land"
                self.land()

            elif 0.2 < distanceVertiBGBM < 0.9 or 0.2 < distanceVertiBMBD < 0.9:
                drone.setHeight(19.2857 * distanceVertiBGBM - 2.3571)
            elif distanceHoriBMOr < -0.3:
                print "left"
                self.left()

            elif distanceHoriBMOr > 0.3:
                print "Right"
                self.right()
            elif pixelHoriBGBD < 9:
                print "forward"
                self.move_forward()

            elif pixelHoriBGBD > 27:
                print "backward"
                self.move_backward()


            else:
                print "Hover"
                self.hover()

        return True




drone = DroneManagment.Drone()

def getMove(zones):
    print "Finding the right move !"
    
    distanceDroneHomme = 5

    
    (barreGauche, barreMilieu, barreDroite) = getBarsName(zones)

    (pixelHoriBGBD,
     distanceHoriBGBD,
     pixelHoriBGBM,
     distanceHoriBGBM,
     pixelHoriBMBD,
     distanceHoriBMBD,
     pixelVertiBGBD,
     distanceVertiBGBD,
     pixelVertiBGBM,
     distanceVertiBGBM,
     pixelVertiBMBD,
     distanceVertiBMBD,
     pixelHoriBMOr,
     distanceHoriBMOr) = getMeasures(barreGauche, barreMilieu, barreDroite)


  
    if len(zones) <= 1 :
        drone.rotateSelf()
    elif len(zones) == 2 :
        if barreMilieu == None :
            return True
        elif barreGauche == None :
            drone.rotateRight()
        else : 
            drone.rotateLeft()
    
    else :
        
        if distanceHoriBGBM < 0.05 and distanceHoriBMBD < 0.05 :
            drone.land()
            
        elif 0.2 < distanceVertiBGBM < 0.9 or 0.2 < distanceVertiBMBD < 0.9 :
            drone.setHeight(19.2857*distanceVertiBGBM - 2.3571)
            
        elif pixelHoriBGBD < 9 :
            drone.move_forward()
                
        elif pixelHoriBGBD > 27 :
            drone.move_backward()
        
        elif distanceHoriBMOr < -0.3 :
            drone.left()
            
        elif distanceHoriBMOr > 0.3 :
            drone.right()
        
        else :
            drone.hover()
    
    return True
    
def getBarsName(zones):

    barreDroite = None
    barreGauche = None
    barreMilieu = None

    for i in range(len(zones)):

        if zones[i].side == False or len(zones) == 1:
            barreMilieu = zones[i]
            barreMilieu.decrisToi("du milieu")

        else:
            if i > 0:

                if zones[i].x > zones[i - 1].x:
                    barreDroite = zones[i]
                    barreDroite.decrisToi("de droite")
                else:
                    barreGauche = zones[i]
                    barreGauche.decrisToi("de gauche")

            else:
                if zones[i].x > zones[i + 1].x:
                    barreDroite = zones[i]
                    barreDroite.decrisToi("de droite")
                else:
                    barreGauche = zones[i]
                    barreGauche.decrisToi("de gauche")

    return (barreGauche, barreMilieu, barreDroite)



def getMeasures(barreGauche, barreMilieu, barreDroite):

    distanceDroneHomme = 5
    pixelHoriBGBD = None
    distanceHoriBGBD = None
    pixelHoriBGBM = None
    distanceHoriBGBM = None
    pixelHoriBMBD = None
    distanceHoriBMBD = None
    pixelVertiBGBD = None
    distanceVertiBGBD = None
    pixelVertiBGBM = None
    distanceVertiBGBM = None
    pixelVertiBMBD = None
    distanceVertiBMBD = None
    pixelHoriBMOr = None
    distanceHoriBMOr = None

    if barreDroite != None and barreGauche != None:
        pixelHoriBGBD = abs(barreDroite.x - barreGauche.x)
        distanceHoriBGBD = 2 * distanceDroneHomme * tan(93 * pi / (180 * 2)) * pixelHoriBGBD / 640

    if barreMilieu != None and barreGauche != None:
        pixelHoriBGBM = abs(barreGauche.x - barreMilieu.x)
        distanceHoriBGBM = 2 * distanceDroneHomme * tan(93 * pi / (180 * 2)) * pixelHoriBGBM / 640

    if barreMilieu != None and barreDroite != None:
        pixelHoriBMBD = abs(barreMilieu.x - barreDroite.x)
        distanceHoriBMBD = 2 * distanceDroneHomme * tan(93 * pi / (180 * 2)) * pixelHoriBMBD / 640

    if barreGauche != None and barreDroite != None:
        pixelVertiBGBD = abs(barreDroite.y - barreGauche.y)
        distanceVertiBGBD = 2 * distanceDroneHomme * tan(76 * pi / (180 * 2)) * pixelVertiBGBD / 480

    if barreMilieu != None and barreGauche != None:
        pixelVertiBGBM = abs(barreGauche.y - barreMilieu.y)
        distanceVertiBGBM = 2 * distanceDroneHomme * tan(76 * pi / (180 * 2)) * pixelVertiBGBM / 480

    if barreMilieu != None and barreDroite != None:
        pixelVertiBMBD = abs(barreMilieu.y - barreDroite.y)
        distanceVertiBMBD = 2 * distanceDroneHomme * tan(76 * pi / (180 * 2)) * pixelVertiBMBD / 480

    if barreMilieu != None:
        pixelHoriBMOr = barreMilieu.x - 320
        if pixelHoriBMOr <= 0:
            distanceHoriBMOr = -2 * distanceDroneHomme * tan(93 * pi / (180 * 2)) * pixelHoriBMOr / 640
        elif pixelHoriBMOr > 0:
            distanceHoriBMOr = 2 * distanceDroneHomme * tan(93 * pi / (180 * 2)) * pixelHoriBMOr / 640

    return (pixelHoriBGBD,
            distanceHoriBGBD,
            pixelHoriBGBM,
            distanceHoriBGBM,
            pixelHoriBMBD,
            distanceHoriBMBD,
            pixelVertiBGBD,
            distanceVertiBGBD,
            pixelVertiBGBM,
            distanceVertiBGBM,
            pixelVertiBMBD,
            distanceVertiBMBD,
            pixelHoriBMOr,
            distanceHoriBMOr)




